### PR TITLE
Preparation for jss-global #324

### PR DESCRIPTION
**Pull request by kof**
This should wait until all plugins, which rely on it are changed as well. Mostly it is just tests.

### CODE REVIEW

**Comment by kof** [comment]
cc @nathanmarks

File: src/RulesFactory.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import RegularRule from './rules/RegularRule'
+import SimpleRule from './rules/SimpleRule'
+import KeyframeRule from './rules/KeyframeRule'
+import ConditionalRule from './rules/ConditionalRule'
+import FontFaceRule from './rules/FontFaceRule'
+
+export default class RulesFactory {
+  constructor() {
+    this.classes = {
```
Comment (nathanmarks) [inline src/RulesFactory.js]: If this is going to be set in the constructor, it should be extendable via an options object. No point otherwise.

File: src/RulesFactory.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+   * Options:
+   *   - `className` pass class name if you to define it manually
+   *
+   * @param {Object} [name]
+   * @param {Object} [style] declarations block
+   * @param {Object} [options] rule options
+   * @return {Object} rule
+   * @api public
+   */
+  get(name, style = {}, options = {}) {
```
Comment (nathanmarks) [inline src/RulesFactory.js]: Wrong method name

File: src/RulesFactory.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import warning from 'warning'
```
Comment (nathanmarks) [inline src/RulesFactory.js]: Why not use flow for this new module?

Comment (kof) [inline src/RulesFactory.js]: will add flow to everything new!

Comment (kof) [inline src/RulesFactory.js]: its a draft

File: src/RulesFactory.js (hunk 2 above)
Comment (kof) [inline src/RulesFactory.js]: what do you mean, bad naming? any suggestions?

File: src/RulesFactory.js (hunk 1 above)
Comment (nathanmarks) [inline src/RulesFactory.js]: Why an option to register new classes but no ability to extend them using the constructor?

File: src/Jss.js [hunk 4]
```diff
@@ -74,22 +75,33 @@ export default class Jss {
...
+    const rule = this.factory.get(name, style, options)
+
+    if (options.applyPlugins !== false) {
+      this.plugins.handleRule(rule)
+    }
+
     return rule
   }
 
+  registerRuleClass(name, cls) {
```
Comment (nathanmarks) [inline src/Jss.js]: This seems like a very low level function to expose here. Part of me feels like the plugin API should be able to handle this use case with atRules all using the same hook.

File: src/Jss.js [hunk 5]
```diff
@@ -17,9 +17,10 @@ export default class Jss {
    * @see .setup()
    */
   constructor(options) {
+    this.version = __VERSION__
     this.sheets = new SheetsRegistry()
-    this.plugins = new PluginsRegistry()
-    this.version = process.env.VERSION
+    this.plugins = new PluginsRegistry(this)
+    this.factory = new RulesFactory()
```
Comment (nathanmarks) [inline src/Jss.js]: This isn't the right property name:

```js
import {create} from 'jss';

const jss = create();

jss.factory // RuleFactory...? Sounds like "JSS Factory" when you read the code which is actually `create()`.
```

File: src/PluginsRegistry.js [hunk 6]
```diff
@@ -6,46 +6,59 @@
...
  */
 export default class PluginsRegistry {
-  registry: Array<Function>;
+  registry: Array<Object>
+  jss: Object
 
-  constructor(): void {
+  constructor(jss: Object): void {
     this.registry = []
+    this.jss = jss
```
Comment (nathanmarks) [inline src/PluginsRegistry.js]: I think that this should be passed into an instance method instead of storing the reference. The plugin registry doesn't need to retain knowledge of the whole `jss` instance IMO.

File: src/RulesFactory.js (hunk 1 above)
Comment (kof) [inline src/RulesFactory.js]: didn't have a use case for it yet.

File: src/Jss.js (hunk 4 above)
Comment (kof) [inline src/Jss.js]: do you have an api in mind?

File: src/Jss.js (hunk 5 above)
Comment (kof) [inline src/Jss.js]: rulesFactory?

File: src/PluginsRegistry.js [hunk 7]
```diff
@@ -6,46 +6,59 @@
...
-    this.runOne(rules)
+  /**
+   * Call all .onRule hooks for passed rule.
+   *
+   * @api public
+   */
+  handleRule(rule: Object): void {
+    for (let index = 0; index < this.registry.length; index++) {
+      const plugin = this.registry[index]
+      if (plugin.onRule) plugin.onRule(rule)
```
Comment (nathanmarks) [inline src/PluginsRegistry.js]: An alternative to conditionally checking every plugin in the registry for the presence of a specific hook every time, is to register the hooks in `.use`. This can be done by having an object holding an array for each hook, eg:

pseudo code:

```js
hooks = {
  onSheet: [],
  onRule: [],
};

use(plugin) {
  Object.keys(plugin).forEach((key) => {
    if (this.hooks[key]) {
      this.hooks[key].push(plugin[key]);
    }
  });
}

handleSheet(sheet) {
  this.hooks.onSheet.forEach((n) => n(sheet));
}
```

File: src/RulesFactory.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    this.classes = {
+      '@charset': SimpleRule,
+      '@import': SimpleRule,
+      '@namespace': SimpleRule,
+      '@keyframes': KeyframeRule,
+      '@media': ConditionalRule,
+      '@supports': ConditionalRule,
+      '@font-face': FontFaceRule
+    }
+    this.regExp = this.buildRegExp()
```
Comment (nathanmarks) [inline src/RulesFactory.js]: Why use a constructor for these use cases? Stylistic preference? I noticed that we have `stage-0` enabled, and class properties are already `stage-2`:

```js
class RulesFactory {
  classes = {
    '@charset': SimpleRule,
  }

  regExp = this.buildRegExp()
}
```

File: src/PluginsRegistry.js (hunk 6 above)
Comment (kof) [inline src/PluginsRegistry.js]: Oh, I could actually pass jss as a second argument to the .use method. This would work for now, not sure though this would work forever, next time it is needed in another method its not any more nice. Also kinda not very nice for the .use method to require this second argument additionally to the plugin. 

I agree that having the jss instance all over the place is not very nice, but I don't see a real good another way. Also which problems can having .jss property bring? Harder to unit test?

File: src/RulesFactory.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  /**
+   * Build regexp for matching rules.
+   *
+   * @return {RegExp}
+   * @api private
+   */
+  buildRegExp() {
```
Comment (nathanmarks) [inline src/RulesFactory.js]: Instead of having a private method like this on the prototype, why not just have it as a function outside of the class?

```js
function buildClassRegExp(classes: Object): RegExp  {
  return new RegExp(`^${Object.keys(classes).join('|')}`)
}
```

File: src/RulesFactory.js (hunk 2 above)
Comment (nathanmarks) [inline src/RulesFactory.js]: Yeah, we shouldn't use `get` for something which **creates**. `get` is the `read` in `CRUD`. This should be `createRule` or `create`.

File: src/PluginsRegistry.js (hunk 7 above)
Comment (kof) [inline src/PluginsRegistry.js]: yep, thats true, do you like this approach in terms of readability or performance?

File: src/StyleSheet.js [hunk 10]
```diff
@@ -20,8 +18,7 @@ import RulesContainer from './RulesContainer'
  * @api public
  */
 export default class StyleSheet {
-  constructor(rules, options) {
-    const named = options.named == null ? true : options.named
+  constructor(styles, options) {
```
Comment (nathanmarks) [inline src/StyleSheet.js]: Why the change to `styles`?

File: src/RulesFactory.js (hunk 9 above)
Comment (kof) [inline src/RulesFactory.js]: can totally be done!

File: src/StyleSheet.js [hunk 11]
```diff
@@ -128,40 +126,40 @@ export default class StyleSheet {
...
+   * Create styles, will render also after stylesheet was rendered the first time.
    *
-   * @param {Object} rules name:style hash.
+   * @param {Object} styles name:style hash.
    * @param {Object} [options]
    * @return {Array} array of added rules
    * @api public
    */
-  addRules(rules, options) {
+  addRules(styles, options) {
```
Comment (nathanmarks) [inline src/StyleSheet.js]: Felt very natural having the first arg to `addRules` named `rules`. What's the change for?

File: src/StyleSheet.js [hunk 12]
```diff
@@ -128,40 +126,40 @@ export default class StyleSheet {
   }
 
   /**
-   * Create rules, will render also after stylesheet was rendered the first time.
+   * Create styles, will render also after stylesheet was rendered the first time.
```
Comment (nathanmarks) [inline src/StyleSheet.js]: Isn't `Create rules` more specific? I'm curious about this change.

File: src/RulesFactory.js (hunk 2 above)
Comment (kof) [inline src/RulesFactory.js]: well its actually a classical factory get method, which is supposed to instantiate and return. I know what you mean, I had it this way at first. But then I looked up the factory pattern and they use `get`.

File: src/rules/RegularRule.js [hunk 13]
```diff
@@ -8,7 +8,7 @@ const {parse, stringify} = JSON
  * @api public
  */
 export default class Rule {
```
Comment (nathanmarks) [inline src/rules/RegularRule.js]: should be `RegularRule`

File: src/StyleSheet.js (hunk 10 above)
Comment (kof) [inline src/StyleSheet.js]: For naming consistency. We call user style definitions styles, they become rules when they are wrapped into a rule class instance.

File: src/StyleSheet.js (hunk 11 above)
Comment (kof) [inline src/StyleSheet.js]: same here, addRules accepts styles object and creates rules out of it, they are not rules at the moment user passes them, seemed better to me as we always communicate it as styles to the user land.

File: src/StyleSheet.js (hunk 12 above)
Comment (kof) [inline src/StyleSheet.js]: hmm actually in this case, its create rules, we take styles and create rules))))

File: src/rules/RegularRule.js (hunk 13 above)
Comment (kof) [inline src/rules/RegularRule.js]: yep

File: src/Jss.js (hunk 4 above)
Comment (nathanmarks) [inline src/Jss.js]: Like, Ideally the plugin API should support the following functionality:

- Plugin can parse the rule name before anything is created
- If it's a global scope selector, the plugin can set the selector as is necessary
- If it is a global scope nested object of rules, the plugin can set the rule type, so `jss` knows it is not a regular CSS rule
- When the nested rules are parsed, the plugin can check if the parent is a globally scoped rule and manipulate the selectors as is necessary

Comment (nathanmarks) [inline src/Jss.js]: I think that this method (`registerRuleClass`) and the injectable `generateClassName` function are unicorns that feel a little bolted on and out of place.

Comment (kof) [inline src/Jss.js]: yep, doesn't feel to me nice as well, will rethink the plugin api to avoid it.

File: src/RulesFactory.js (hunk 8 above)
Comment (nathanmarks) [inline src/RulesFactory.js]: Those are instance properties not static.

File: src/RulesFactory.js (hunk 2 above)
Comment (kof) [inline src/RulesFactory.js]: I think it implies that an instance is not always created, it might be cached and just returned.

Comment (nathanmarks) [inline src/RulesFactory.js]: I've seen create or createxxxx used in loads of factory patterns, including textbook examples etc. But I digress... Up to you.

File: src/Jss.js (hunk 5 above)
Comment (nathanmarks) [inline src/Jss.js]: That works.

File: src/PluginsRegistry.js (hunk 7 above)
Comment (nathanmarks) [inline src/PluginsRegistry.js]: That's a good question. It depends on a number of variables including the size of the application. I need to test to get specifics but I'm interested in finding out.

**Comment by kof** [comment]
I think it makes sense, because different hooks are called at different times, Its nicer to have an overview like this. Perf difference will be minimal but yeah, to call the specific hook iterator needs to iterate less, but as the number is just a bunch, it won't matter that much.

File: src/RulesFactory.js (hunk 2 above)
Comment (kof) [inline src/RulesFactory.js]: ok changed.

File: src/plugins.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  '@font-face': FontFaceRule
+}
+
+/**
+ * Generate plugins which will register all rules.
+ */
+const plugins = Object.keys(classes).map((key) => {
+  const re = new RegExp(`^${key}`)
+  const onCreate = (name, style, options) => (
+    re.test(name) ? new classes[key](name, style, options) : null
```
Comment (nathanmarks) [inline src/plugins.js]: Why not just, `name.indexOf(key) === 0`? Don't have to create the regexp every time that way

**Comment by kof** [comment]
To be sure I just made some benchmarking

https://jsperf.com/indexof-vs-substr-vs-regex-at-the-beginning-3 <https://jsperf.com/indexof-vs-substr-vs-regex-at-the-beginning-3>

On Chrome: regex and indexof perform the same, substr. is twice faster. I assume my bench is not good enough and chrome tricks it out with some optiomizations

Safari, Firefox, IE11 - regex is the fastest.

What you probably oversaw is that the regex is created just once for each plugin, onCreate doesn't generate the regex each time. Hot part is only onCreate, plugin genearation happens just once for each plugin.

**Comment by nathanmarks** [comment]
@kof Yeah, I totally overlooked that. Thanks for doing the pef.

I did some additional testing myself using a smaller string. In Firefox and Safari, the regex was a lot faster. 

I was surprised. I didn't know that it was faster for small strings. TIL.

**Comment by kof** [comment]
merged
